### DISCUSSION TITLE
"Quickstart: Running Cognita Locally" sample test requires installing `additional.requirements.txt`

### DISCUSSION THREAD

**Discussion by avibathula** [truefoundry/cognita#163] [Q&A] [answered]
I tried to follow the instructions of `README.md`'s "Quickstart: Running Cognita Locally" section and ran into following errors when running `python -m local.ingest` command

`ModuleNotFoundError: No module named 'torch'`
    -  ***looks like requires the sample test requires installing `additional.requirements.txt`***
    - The error went away after running `pip install -r additional.requirements.txt`

**Comment by chiragjn** [comment, marked as answer]
We have now split up requirements according to major components and updated our Readme. We would appreciate your feedback on this

**Reply by avibathula** [reply]
It was more than a week back and now I do not even remember what I was doing :( - will have to start from scratch and retry and see.

**Reply by avibathula** [reply]
Thank you @chiragjn

**THE GOOD**: Documentation is more clear now.

**THE NOT SO GOOD :)** : `pip install -r backend/requirements.txt` is failing with

```
ERROR: Cannot install -r backend/requirements.txt (line 2), -r backend/requirements.txt (line 24), -r backend/requirements.txt (line 3), -r backend/requirements.txt (line 31), -r backend/requirements.txt (line 5), -r backend/requirements.txt (line 7), requests==2.32.2, truefoundry and truefoundry[ml]==0.2.1 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested requests==2.32.2
    langchain 0.1.9 depends on requests<3 and >=2
    langchain-community 0.0.24 depends on requests<3 and >=2
    langchain-core 0.1.30 depends on requests<3 and >=2
    tiktoken 0.6.0 depends on requests>=2.26.0
    truefoundry 0.2.1 depends on requests<3.0.0 and >=2.31.0
    huggingface-hub 0.20.3 depends on requests
    transformers 4.40.2 depends on requests
    truefoundry[ml] 0.2.1 depends on requests<3.0.0 and >=2.31.0
    servicefoundry 0.10.11 depends on requests<2.32.0 and >=2.23.0

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

For full log, see [pip-install-backend-requirements-failure.txt](https://github.com/truefoundry/cognita/files/15492013/pip-install-backend-requirements-failure.txt)

---

My local repo's top git log entry is 

```
**commit 5932ae9791cfa7c47437dc5b4e4652daca87f349 (HEAD -> main, origin/main, origin/HEAD)
Merge: 33b6651 1ec08e5
Author: Chirag Jain <jain.chirag925@gmail.com>
Date:   Thu May 30 00:16:13 2024 +0530

    Merge pull request #158 from truefoundry/dependabot/pip/backend/fastapi-0.109.1

    Bump fastapi from 0.104.1 to 0.109.1 in /backend**
    ```

**Reply by avibathula** [reply]
@chiragjn - I was able to resolve the issue using `pip-compile` - find attached the new/updated requirements.txt file
[requirements.txt](https://github.com/truefoundry/cognita/files/15493305/requirements.txt) that addresses the build failure.

One problem with this is that the requirements.txt file gets flattened out and readers cannot distinguish between direct dependencies and indirect-dependencies. I feel Cognita has reached a place where it needs to graduate to using poetry : https://github.com/truefoundry/cognita/issues/190

CC: @agutgutia1991

**Comment by agutgutia1991** [comment]
Apologies for the delay in reply. Please do give it a try again and we will
ensure we get back soon on this.
Could you share your email? I will add you to an internal Cognita slack
where we can help resolve issues faster till we get a community
slack/discord live.
Thanks,
Anuraag

**Reply by avibathula** [reply]
I just pinged you on LinkedIn.
